Sources/PubNub/EventEngine/Subscribe/SubscribeTransition.swift

```diff
@@ -0,0 +1,325 @@
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+class SubscribeTransition: TransitionProtocol {
+  typealias State = (any SubscribeState)
+  typealias Event = Subscribe.Event
+  typealias Invocation = Subscribe.Invocation
+  
+  private func willExit(from state: State) -> [EffectInvocation<Invocation>] {
+    switch state {
+    case is Subscribe.HandshakingState:
+      return [.cancel(id: Invocation.ID.HandshakeRequest)]
+    case is Subscribe.HandshakeReconnectingState:
+      return [.cancel(id: Invocation.ID.HandshakeReconnect)]
+    case is Subscribe.ReceivingState:
+      return [.cancel(id: Invocation.ID.ReceiveMessages)]
+    case is Subscribe.ReceiveReconnectingState:
+      return [.cancel(id: Invocation.ID.ReceiveReconnect)]
+    default:
+      return []
+    }
+  }
+  
+  func transition(
+    from state: State,
+    event: Event
+  ) -> TransitionResult<State, Invocation> {
+    var results: TransitionResult<State, Invocation>
+    
+    switch event {
+    case .reconnect:
+      switch state {
+      case let state as any CursorState:
+        results = setReceivingState(from: state, cursor: state.cursor)
+      default:
+        results = setHandshakingState(from: state)
+      }
+    case .handshakeSucceess(let cursor):
+      results = setReceivingState(from: state, cursor: cursor)
+    case .handshakeFailure(let error):
+      results = setHandshakeReconnectingState(from: state, error: error)
+    case .handshakeReconnectSuccess(let cursor):
+      results = setReceivingState(from: state, cursor: cursor)
+    case .handshakeReconnectFailure(let error):
+      results = setHandshakeReconnectingState(from: state, error: error)
+    case .handshakeReconnectGiveUp(let error):
+      results = setHandshakeFailedState(from: state, error: error)
+    case .receiveSuccess(let cursor, let messages):
+      results = setReceivingState(from: state, cursor: cursor, messages: messages)
+    case .receiveFailure(let error):
+      results = setReceiveReconnectingState(from: state, error: error)
+    case .receiveReconnectSuccess(let cursor, let messages):
+      results = setReceivingState(from: state, cursor: cursor, messages: messages)
+    case .receiveReconnectFailure(let error):
+      results = setReceiveReconnectingState(from: state, error: error)
+    case .receiveReconnectGiveUp(let error):
+      results = setReceiveFailedState(from: state, error: error)
+    case .subscriptionChanged(let channels, let groups):
+      results = onSubscriptionChanged(from: state, newChannels: channels, newGroups: groups)
+    case .subscriptionRestored(let channels, let groups, let cursor):
+      results = onSubscriptionRestored(from: state, newChannels: channels, newGroups: groups, cursor: cursor)
+    case .disconnect:
+      results = setStoppedState(from: state)
+    }
```

Comment:
Am I wrong, or with this implementation, there is no check on whether the current state can handle an event or not?
So, for example, what will happen if state machine is in `ReceivingStopped` and will receive `handshakeFailure` event?

Reply:
Not all of them, but some helper methods listed here like [setHandshakeReconnectingState](https://github.com/pubnub/swift/blob/feat/event-engine-effects/Sources/PubNub/EventEngine/Subscribe/SubscribeTransition.swift#L168-L174) contains `guard` statements in order to check whether the transition should be allowed or not. I could use the same approach for all functions as well. Also, your comment proved to me those functions' names might be extremely confusing. I think it might be better to rename them like `trySetHandshakeReconnectState` or whatever else you come up with.  

I was thinking about the use case with `ReceivingStopped` and `handshakeFailure` you mentioned above. We have roughly 15 events, but only a few of them can be passed to EE due to customer's actions like disconnecting, adding/removing new channels, or restoring subscription. Most of the events passed to EE are the results from our Effects, so my idea was to write unit tests to ensure that every Effect dispatches the expected `Event` back to EE when succeeded/failed. I wanted to repeat the same approach for the hypothetical class like `SubscriptionManager` that holds a reference to EE. Such a class should contain a set of unit tests to verify whether expected events are dispatched to EE as a result of customers' actions like subscribing/unsubscribing/restoring subscription, etc. To sum up, I wanted to rely on unit tests as a safety net in order to prevent writing extra assertions in source code. Do you have any other alternatives/ideas?